### ISSUE TITLE
FR: Pushing and fetching arbitrary refs

### ISSUE THREAD

**Issue by ilyagr** [martinvonz/jj#4388]
**Is your feature request related to a problem? Please describe.**
Currently, there is no way to use `jj` to push or fetch an arbitrary ref from the remote. This causes problems for e.g. Gerrit as [discussed on Discord](https://discord.com/channels/968932220549103686/1281020382249357423/1281020382249357423).

On one hand, this reinforces the desire for https://github.com/martinvonz/jj/issues/4387, but separately it would be nice to support something for the general case.

**Describe the solution you'd like**
I can think of 2 possible solutions:

1. Re-design our branches (soon to be called bookmarks) to support tracking an arbitrary ref on a remote. This probably would imply having to support tracking different refs on different remotes, tracking remote branches of a different name, and perhaps tracking multiple refs per remote.
2. Set up simple commands to push and fetch arbitrary refs unsafely, either permanently or until we can implement 1 or think of a better solution.

Here's a possible interface for option 2:

- `jj git fetch-ref [--remote REMOTE] REF_NAME [branch_name]` would fetch an arbitrary ref and optionally create a branch at the newly fetched commit for easier reference later. This new local branch wouldn't track anything. It'd just be a bookmark :).
- `jj git push-ref <--force|--expected-at COMMIT_ID> [--remote REMOTE] REVISION REF_NAME`

Both 1 and 2 have disadvantages. 1 adds complexity to the UI and the model. I'm not sure whether or not 2. is better than telling people to co-locate the repo and use `git fetch/git push`.

In any case, I'm curious what people think should be done here.

**Comment by yuja** [comment]
Some previous discussions:

> 1. Re-design our branches (soon to be called bookmarks) to support tracking an arbitrary ref on a remote.

https://github.com/martinvonz/jj/issues/3235

> 2. Set up simple commands to push and fetch arbitrary refs

https://github.com/martinvonz/jj/pull/2098#issuecomment-1682522251

For 2, we might want to introduce command namespace for plumbing commands so `jj git fet[TAB]` wouldn't stop at `jj git fetch(-refs)?`.

**Comment by emilazy** [comment]
This is also nice for GitHub (refs like `pull/123/head`).

**Comment by ilyagr** [comment]
Thanks, I forgot about those. 

For a subcommand, `jj git ref fetch` could work.

**Comment by Minion3665** [comment]
Heya! I was the one who originally prompted the discussion on Discord - I have a few thoughts to put here too :)

> Re-design our branches (soon to be called bookmarks) to support tracking an arbitrary ref on a remote

This seems interesting to me, although I note that it wouldn't quite solve the gerrit/agit/radicale/idk use-case in a nice way, at least for pushes. That's because so-called "magic refs" do weird things on the server - you always push to `refs/for/main`, etc., and you can't pull them back down from the same place. I'm not sure how much software uses magic refs like this; my assumption is that a small but certainly nonzero amount does.

That said: my use-case only particularly cares about pulls, and I generally think that they are more widely useful. The [`refs/pull/#/head`](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/checking-out-pull-requests-locally#modifying-an-inactive-pull-request-locally) mentioned earlier are a great example of this - GitHub is really widely used!

> 1 adds complexity to the UI and the model

This is probably true, and is a shame. No matter how hard we try to make it so refs aren't exposed unless you're using the feature (hide the refs/foo/... prefix if foo is heads, etc.), probably it at least makes the documentation more confusing if you don't know what a ref is.

> I'm not sure whether or not 2. is better than telling people to co-locate the repo and use `git fetch/git push`

To me it would be

- I'm using nix with flakes, nix has a thing where files that are untracked aren't seen at all, which has left me very confused about why my code isn't doing anything more times that I would like to admit. jj "solves" this by not being git, and therefore not having this oddity. More generally: some tools integrate poorly with Git. It can be a disadvantage to have a colocated repository.

Additionally, if I do want to push to arbitrary refs...

- I want to refer to things in the way I am doing with `jj`. I often `jj edit change_id`, leaving my changes in git uncommited. It's a bit more friction to `jj show @`, copy the commit ID, and go over to git, given there's no HEAD reference for that.

I'm sure that I could write a hacky bash function to workaround these issues (using the underlying git repo and a lot of duct tape) and go on my way... call it `jj.git.ref` or something, but I'm not particularly enthusiastic...

**Comment by ilyagr** [comment]
@jennings suggested using `--named` for this on Discord. I'm not sure I like that, but perhaps `--ref` would work. Copying the examples he used with this minor modification, it could look like this:

```
jj git fetch --ref pr123=refs/pull/123/head
jj git push --ref ref/for/main=xyz
```

(I'm not claiming that this is necessarily better than the syntax old me suggested, or a syntax anyone else suggested. I just wanted to write the fresh idea down before I forget.)

**Comment by 9999years** [comment]
I would like to register my interest in being able to fetch arbitrary _commits_ as well as refs. This is useful for backports or cherry-picks, especially on GitHub where repo contains all of its forks in the backing storage, so you can do `git fetch my_github_remote some_commit_hash` to fetch a commit from an arbitrary fork (see #7369).

**Comment by djmitche** [comment]
I started to take a look at the code around fetching. I'd be very interested in implementing a solution (with some assistance), but it seems that there's not yet consensus on what that should look like. Maybe that offer makes it a bit more attractive to pick an approach? The suggestion two comments before this one seems workable. AIUI that would essentially set up tracking for a local name to a remote ref, rather than just to a remote branch.

**Comment by Minion3665** [comment]
@djmitche as I understand it, that wouldn't set up tracking - rather `jj git fetch --ref pr123=refs/pull/123/head` would create a new local bookmark, `pr123` that was at whatever commit `refs/pull/123/head` referenced - but `jj git fetch` would not fetch `refs/pull/123/head` again if that changed...

...similarly for `jj git push --ref refs/for/main=xyz`, whatever single commit was referenced by revset `xyz` would be pushed to `refs/for/main` but - if `xyz` was a local bookmark - running `jj git push` after updating `xyz` wouldn't make it push to `refs/for/main` again

Tracking could be cool but (quoting [myself from earlier in this thread](https://github.com/jj-vcs/jj/issues/4388#issuecomment-2336517216))

> This seems interesting to me, although I note that it wouldn't quite solve the gerrit/agit/radicle/idk use-case in a nice way, at least for pushes. That's because **so-called "magic refs" do weird things on the server - you always push to refs/for/main, etc., and you can't pull them back down from the same place**. I'm not sure how much software uses magic refs like this; my assumption is that a small but certainly nonzero amount does

(emphasis mine... from now, not then)

---

I do really like the suggestion from ~~2~~ 3 comments ago too - I'd be very happy if we got something like that

**Comment by bnjmnt4n** [comment]
I actually have some preliminary code for fetching refs here: https://github.com/jj-vcs/jj/pull/5630. It implements the command `jj git ref fetch <REF>`. The other suggestion is `jj git fetch --ref newbookmark=ref`, which could be useful because it allows fetching multiple refs at a time. However, the downside of this option structure as compared to `jj git ref fetch` is that it doesn’t allow the commits to be fetched anonymously, i.e. without creating a local bookmark.

(I’m not sure how useful anonymous commits are. I would assume that if someone is fetching a specific ref instead of a typical bookmark from the remote, the commit being fetched is typically transient, and so creating a bookmark would be redundant.)

**Comment by Minion3665** [comment]
> I actually have some preliminary code for fetching refs here: [#5630](https://github.com/jj-vcs/jj/pull/5630). It implements the command `jj git ref fetch <REF>`. The other suggestion is `jj git fetch --ref newbookmark=ref`, which could be useful because it allows fetching multiple refs at a time. However, the downside of this option structure as compared to `jj git ref fetch` is that it doesn’t allow the commits to be fetched anonymously, i.e. without creating a local bookmark.
> 
> (I’m not sure how useful anonymous commits are. I would assume that if someone is fetching a specific ref instead of a typical bookmark from the remote, the commit being fetched is typically transient, and so creating a bookmark would be redundant.)

hmmm, could we `jj git fetch --ref [newbookmark=]ref` to allow extending this syntax to anonymous commits ... i.e. `jj git fetch --ref refs/pull/123/head` would fetch the ref and say something like

<img width="1469" height="95" alt="Image" src="https://github.com/user-attachments/assets/b51779fb-2bd9-4f3b-b7b3-33a4c3fd0d83" />

whereas you could also `jj git fetch --ref pr123=refs/pull/123/head` to set the local `pr123` bookmark to the fetched commit (in this case `xmysvlzo`)

**Comment by bnjmnt4n** [comment]
One potential issue is that the `=` character can be part of the ref name, so the `[opt_bookmark=]ref` syntax could be ambiguous.

**Comment by yuja** [comment]
Since `jj git fetch`/`push --ref` would be incompatible with the other options, I think it's better to add separate commands for low-level fetch/push. We'll need force push flag, etc.

> One potential issue is that the `=` character can be part of the ref name,

We could parse the right-hand side as string pattern, but I'm not sure if we should overload `bookmark=<single-ref>` and `<ref-with-glob>`.

**Comment by Minion3665** [comment]
> We'll need force push flag, etc.

This is true and is a great argument for separate commands - I don't want a `--force` flag on my normal push command so I'm convinced in the betterness of `jj git ref fetch/push` :)

**Comment by oliverpool** [comment]
Maybe the pushing and fetching aspects should be handled separately? (I am especially interested in fetching, for local PR-review from GitHub, Forgejo, GitLab...).

`jj git fetch --ref refs/pull/123/head` looks good and for local bookmark, maybe `:` could be used as separator? (since [git forbids it in ref names](https://git-scm.com/docs/git-check-ref-format#:~:text=colon%20%3A%20anywhere))

So `jj git fetch --ref pr123:refs/pull/123/head` would set the local `pr123` bookmark to the fetched commit.

This would also allow multiple fetching at once.

Pushing seems to be a complete different challenge (`--force` flag for instance) and might be better served by a dedicated subcommand.

**Comment by tingerrr** [comment]
Does pulling a ref require us to set a bookmark? I could see this as a nice analogy to `jj git push --named <bookmark>=<revision>` (with `=` instead of `:` for consistency), but in theory fetching and then importing it as a simple anonymous change should be fine too, right?

Does git do anything fancy when pulling a ref? Does it create the ref in the local git object store too? Does it require you to assign a branch name?

I agree with the sentiment of implementing push at a later date, but I think it should not live under a different command, but this likely depends on whatever plumbing is necessary in git to do this.

**Comment by oliverpool** [comment]
> Does git do anything fancy when pulling a ref? 

I don't think so. Here is what I see, when trying with `git fetch origin pull/10484/head` on some repo:
```
remote: ...
Receiving objects: ...
Resolving deltas: ...
From ...
 * branch                  refs/pull/10484/head -> FETCH_HEAD
```

> Does it create the ref in the local git object store too?

`git show-ref` does not list this newly fetched commit, so I would say no.

> Does it require you to assign a branch name?

No, see above. But without a branch name, it is harder to do anything with the fetched objects.

When fetching 2 PRs and a branch at once, the `.git/FETCH_HEAD` file looks like this (overwritten on next fetch):

```
83afc53f006fe037b15b45f4537542f979e3323a        'refs/pull/10487/head' of https://...
1c73d5dbbe719a4fe3266aed66469d7e2c54b7f5        'refs/pull/10484/head' of https://...
b90eda6ed70ceff415b4d3b856b0eddd8089d721        branch '...' of https://...
```

Passing the [`--porcelain`](https://git-scm.com/docs/git-fetch#Documentation/git-fetch.txt---porcelain) flag, allows to get the list of shas, in the same order as requested, on stdout (specific sha can be retrieved the same way, see #7369) :
```
* 0000000000000000000000000000000000000000 83afc53f006fe037b15b45f4537542f979e3323a FETCH_HEAD
* 0000000000000000000000000000000000000000 1c73d5dbbe719a4fe3266aed66469d7e2c54b7f5 FETCH_HEAD
* 0000000000000000000000000000000000000000 b90eda6ed70ceff415b4d3b856b0eddd8089d721 FETCH_HEAD
```

**Comment by tomfitzhenry** [comment]
FYI, I think pushing/fetching arbitrary refs is needed for gittuf integration. gittuf is a security policy layer for commit signing in Git.

I've raised https://github.com/gittuf/gittuf/discussions/1271 on gittuf side.

**Comment by mahlzahn** [comment]
@oliverpool:

> `jj git fetch --ref refs/pull/123/head` looks good and for local bookmark, maybe `:` could be used as separator? (since [git forbids it in ref names](https://git-scm.com/docs/git-check-ref-format#:~:text=colon%20%3A%20anywhere))
> 
> So `jj git fetch --ref pr123:refs/pull/123/head` would set the local `pr123` bookmark to the fetched commit.

Not a user of jj, but just to note that `git fetch` does the opposite:
```
git fetch -u origin refs/pull/123/head:pr123
```
gets the remote ref and creates a new branch pr123.

**Comment by oliverpool** [comment]
#9643 by @joshka propose to add `jj git ref fetch`, `jj git ref list`, and `jj git ref forget`

**Comment by neild** [comment]
I've been eying this issue with interest, since I'd like a better way to fetch Gerrit patch sets (which exist as non-bookmark/tag references with names like `refs/changes/01/818201/2`).

Playing around with possible implementations (mostly for my own amusement), I came up with #10225. I wonder if this might be a reasonable alternative to the one-shot ref fetch path in #9993?

The approach I took was to extend `GitRefKind` (which enumerates the possible types of remote reference) to add an "Other" ref in addition to Bookmark and Tag. This means no need for a new fetch path, and the possibility of revsets that reference a non-bookmark/tag ref.

Curious whether this seems like a useful direction to pursue?

(I'm mainly interested in the internal APIs here, not the CLI syntax.)
